pulsar-client-cpp/lib/ConnectionPool.h

```diff
@@ -59,7 +61,7 @@ class PULSAR_PUBLIC ConnectionPool {
     ClientConfiguration clientConfiguration_;
     ExecutorServiceProviderPtr executorProvider_;
     AuthenticationPtr authentication_;
-    typedef std::map<std::string, ClientConnectionWeakPtr> PoolMap;
+    typedef std::map<std::string, ClientConnectionPtr> PoolMap;
```

Comment:
What is the reason for changing the `weak_ptr` into a `shared_ptr`? 

This could also potentially have side effects, like not destructing the connections (that are already closed) while the pool is active. 

As I understand this change, the reason is to iterate through the map in the `~ConnectionPool` and force calling `ClientConnectionPtr::close()`. Couldn't we just do the same with the `weak_ptr` and calling `lock()` while iterating through the map and closing the ptr that are still valid?

Reply:
> What is the reason for changing the `weak_ptr` into a `shared_ptr`? 

While iterating the pool map, I am worried that the iterator would be broken when `ClientConnection` is destructed.

> Couldn't we just do the same with the `weak_ptr` and calling `lock()` while iterating through the map and closing the ptr that are still valid?

I will try to fix it like that.

Reply:
> While iterating the pool map, I am worried that the iterator would be broken when ClientConnection is destructed.

The iterator itself is on a map in `ConnectionPool` so that will still be always safe to use here.

The `ClientConnectionWeakPtr` can, of course, be possibly already destroyed, though the `lock()` operation will attempt to acquire a ref count on the object and get a `shared_ptr`. The resulting `shared_ptr` has to be checked if it's valid, and if it is, it will be kept alive by the shared_ptr.